torch/csrc/autograd/function.h

```diff
@@ -369,6 +370,12 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
   /// Returns the name of the dynamic type of the function, for debugging.
   virtual std::string name() const;
 
+  /// The difference between functions `should_compute_output` and
+  /// `task_should_compute_output`: `should_compute_output` is called
+  /// when constructing the node because the node itself is irreverent to the
+  /// graph_task, `task_should_compute_output` is called during the node
+  /// execution.
+  ///
```

Comment:
> We could avoid any codegen changes (and bring these improvements by default for anyone using should_compute_output) by using the global thread_local graph_task variable. Instead of having to pass it as an argument (and having to create a new method for Node) we can update the existing should_compute_output(idx) to check the thread_local graph_task. Currently the thread_local graph_task variable is declared in engine.cpp, but we can lift that into engine.h.

yes, however there are a few problems
1. when constructing the node at [gen_variable_type.py#L958](https://github.com/pytorch/pytorch/blob/d13c40d92281eca1e1901c86baa4e8dc4ff64922/tools/autograd/gen_variable_type.py#L958), we should not check graph_task, because the node itself is irrelevant to the task.
2. I was assuming graph_task is empty when constructing nodes, but it seems not true if we also do double backward. There are a few tests that failed if I check graph_task while constructing nodes.
3. `engine.h` include `function.h`, and `GraphTask` is declared at `engine.h`, if we move `GraphTask` into a single header and include it in `function.h`, this will introduce errors like `This change adds a dependency on all pytorch operators, meaning the       file will need to be re-compiled every time an operator is changed or added.   Consider including a specific operator from <ATen/ops/{my_operator}.h> and     see NOTE [TORCH_ASSERT_ONLY_METHOD_OPERATORS].`, which is bad. That's why I only moved `ExecInfo` into a single header, this is minimal and sufficient to achieve the same goal.

Reply:
Good point. If we'd like to avoid ANY bc-breaking changes to `should_compute_output`, it does seem like we should not update it or it would fail in double backward because we would otherwise be unable to distinguish between a call made from Functions.cpp or VariableTypeX.cpp. 

I am a little concerned about the naming `task_should_compute_output` though since we want to expose that as public API for cpp custom Function users and "task" is more of a concept internal to the engine. Perhaps we want to bikeshed on that a bit more? Its kind of annoying that `should_compute_output` is already taken to mean "should build graph for output". One idea is: if we don't think any folks are using `should_compute_output` (outside of codegen) for the purposes of building the graph, then maybe we are okay to break that particular use case, i.e. actually rename the current implementation of `should_compute_output` to `should_build_graph` and replace usages of that in VariableType codegen.

So now it seems that whether there is still benefit to using TLS now depends on whether we are okay with breaking the use case of  `should_compute_output` for graph construction to allow for usages of it during backward to benefit from the speed up without code changes. (cc @albanD for if this OK).

Regarding (3), I don't completely understand why we should encounter that message here because engine.cpp or graph_task shouldn't really depend on any codegen'd operators. I checked that engine.h depends on TensorBody.h (transitively through Tensor.h) and that we still build succesfully after removing that. Perhaps that will resolve this issue?

Reply:
@soulitzer Thanks for the detailed explanation!

I like `should_build_graph` and have changed it accordingly. @albanD this PR introduces breaking changes to `should_compute_output ` function. The original `should_compute_output` is renamed to `should_build_graph` which is only called at codegen. Let me know if you have any concerns, and I could revert it accordingly.

Regarding (3), I debugged again. It is the `input_buffer.h` that caused the problem, GraphTask needs the full implementation of InputBuffer,

https://github.com/pytorch/pytorch/blob/6d7b7615b1f9d9b7853f691260a399d61c4d7ee7/torch/csrc/autograd/engine.h#L68

and `input_buffer.h` includes unnecessary <ATen/ATen.h>..
https://github.com/pytorch/pytorch/blob/6d7b7615b1f9d9b7853f691260a399d61c4d7ee7/torch/csrc/autograd/input_buffer.h#L8

Everything has been fixed now, and let me know if you have any other suggestions.